pkg/apis/storage/validation/validation.go

```diff
@@ -430,7 +430,6 @@ func ValidateCSIDriverUpdate(new, old *storage.CSIDriver) field.ErrorList {
 	allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(new.Spec.FSGroupPolicy, old.Spec.FSGroupPolicy, field.NewPath("spec", "fsGroupPolicy"))...)
 	allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(new.Spec.PodInfoOnMount, old.Spec.PodInfoOnMount, field.NewPath("spec", "podInfoOnMount"))...)
 	allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(new.Spec.VolumeLifecycleModes, old.Spec.VolumeLifecycleModes, field.NewPath("spec", "volumeLifecycleModes"))...)
-	allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(new.Spec.StorageCapacity, old.Spec.StorageCapacity, field.NewPath("spec", "storageCapacity"))...)
```

Comment:
The scheduler may also need updating similar to https://github.com/kubernetes/kubernetes/pull/100026#event-4790034661. Would a pod that failed scheduling be schedulable after this object is added/removed/updated?

Reply:
> The scheduler may also need updating similar to #100026 (comment). 

That link just takes me to the PR. Are you suggesting that the scheduler needs to watch CSIDriver objects and attempt to reschedule pods when those objects change?

> Would a pod that failed scheduling be schedulable after this object is added/removed/updated?

Yes. Scheduling might be blocked when the value was "true" and there is no capacity. When the value becomes "false", scheduling could proceed.

But this seems a bit unlikely. The main use case for mutating CSIDriver was to start with the "false" and later switch to "true".

My expectation was that pod scheduling is going to be retried. That seems sufficient for the less common other direction.

Reply:
Even though it may not be common, it would be good to handle for cases such as rollback/disablement. Similarly, the scheduler probably needs to handle CSIStorageCapacity objects as well. Anyway, I think this is orthogonal to the api change and can be investigated separately.

Reply:
But pod scheduling for `true -> false` is handled, isn't it? It just doesn't happen immediately after the change.

Reply:
cc @ahg-g 

My old understanding is that we would move the pod back to the scheduling queue after some period of time, but I'm not sure if that's still the case anymore and if it's all event based.

Reply:
if this is dropping the immutability validation, it needs to start validating the new value is valid (calling validateStorageCapacity instead), right?

Reply:
> My old understanding is that we would move the pod back to the scheduling queue after some period of time, but I'm not sure if that's still the case anymore and if it's all event based.

If we're comparing this to the scheduler's current behavior, I suspect it would behave the same as it currently does if you deleted the CSIDriver and recreated it with the setting you wanted. Whether that immediately retries to schedule pods that were previously unschedulable, I don't know (cc @alculquicondor @ahg-g to answer that question). It seems like speeding up scheduler requeuing of unschedulable pods if a relevant reason they were unschedulable changes could be done separately from this PR (but would be worth tracking in an issue)

Reply:
> The main use case for mutating CSIDriver was to start with the "false" and later switch to "true".

that makes sense to me, especially since it's a new field that defaults to false, so all pre-existing CSIDriver objects will be set to `false`, and any that want to add support for this feature have to change the value to `true`.

Reply:
We are working towards scheduling be purely event based. But we still have a routine that puts Pods back into the scheduling queue after 1min.
Either way, the volumebinding plugin already knows to react to CSIDriver updates https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/volumebinding/volume_binding.go#L116

> if this is dropping the immutability validation, it needs to start validating the new value is valid (calling validateStorageCapacity instead), right?

Shouldn't we just call `ValidateCSIDriver` on the new object to be future proof against other mutability changes?

Reply:
PR for the KEP: https://github.com/kubernetes/enhancements/pull/2854

I'm not sure whether we have specific documentation for `CSIDriver` besides the auto-generated API documentation which will be updated as part of this PR.

Reply:
> It seems like speeding up scheduler requeuing of unschedulable pods if a relevant reason they were unschedulable changes could be done separately from this PR (but would be worth tracking in an issue)

https://github.com/kubernetes/kubernetes/issues/104250